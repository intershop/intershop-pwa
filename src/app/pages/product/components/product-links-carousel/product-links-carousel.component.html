<div *ngIf="links.productSKUs.length" class="product-list-container">
  <h2>{{ productLinkTitle }}</h2>

  <div class="product-list">
    <swiper [config]="swiperConfig">
      <div *ngFor="let sku of links.productSKUs" class="swiper-slide">
        <ish-product-item-container [productSku]="sku"></ish-product-item-container>
      </div>
    </swiper>
  </div>
</div>
