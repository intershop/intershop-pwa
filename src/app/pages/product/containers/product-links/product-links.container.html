<pre>{{ links$ | async | json }}</pre>
<ng-container *ngIf="links$ | async as links">
  <ng-container *ngIf="links.upselling.productSKUs.length > 0">
    <ish-product-links-list
      [links]="links.upselling"
      [productLinkTitle]="'product.product_links.upselling.title' | translate"
    ></ish-product-links-list>
  </ng-container>
  <ng-container *ngIf="links.crossselling.productSKUs.length > 0">
    <ish-product-links-carousel
      [links]="links.crossselling"
      [productLinkTitle]="'product.product_links.crossselling.title' | translate"
    ></ish-product-links-carousel>
  </ng-container>
</ng-container>
