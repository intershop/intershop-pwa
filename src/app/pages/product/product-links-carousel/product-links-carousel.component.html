<div *ngIf="links.products.length" class="product-list-container">
  <h2>{{ productLinkTitle }}</h2>

  <div class="product-list">
    <swiper [config]="swiperConfig">
      <div *ngFor="let sku of links.products" class="swiper-slide">
        <ng-template swiperSlide>
          <ish-product-item ishProductContext [sku]="sku"></ish-product-item>
        </ng-template>
      </div>
    </swiper>
    <div class="swiper-pagination"></div>
  </div>
</div>
