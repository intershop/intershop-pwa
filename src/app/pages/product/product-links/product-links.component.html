<ng-container *ngIf="links$ | async as links">
  <ng-container *ngIf="links.upselling?.products?.length">
    <ish-product-links-list
      [links]="links.upselling"
      [productLinkTitle]="'product.product_links.upselling.title' | translate"
    ></ish-product-links-list>
  </ng-container>
  <ng-container *ngIf="links.spareparts?.products?.length">
    <ish-product-links-list
      [links]="links.spareparts"
      [productLinkTitle]="'product.product_links.spareparts.title' | translate"
    ></ish-product-links-list>
  </ng-container>
  <ng-container *ngIf="links.crossselling?.products?.length">
    <ish-product-links-carousel
      [links]="links.crossselling"
      [productLinkTitle]="'product.product_links.crossselling.title' | translate"
    ></ish-product-links-carousel>
  </ng-container>
  <ng-container *ngIf="links.accessory?.products?.length">
    <ish-product-links-carousel
      [links]="links.accessory"
      [productLinkTitle]="'product.product_links.accessory.title' | translate"
    ></ish-product-links-carousel>
  </ng-container>
  <ng-container *ngIf="links.differentorderunit?.products?.length">
    <ish-product-links-carousel
      [links]="links.differentorderunit"
      [productLinkTitle]="'product.product_links.differentorderunit.title' | translate"
    ></ish-product-links-carousel>
  </ng-container>
  <ng-container *ngIf="links.followup?.products?.length">
    <ish-product-links-carousel
      [links]="links.followup"
      [productLinkTitle]="'product.product_links.followup.title' | translate"
    ></ish-product-links-carousel>
  </ng-container>
  <ng-container *ngIf="links.other?.products?.length">
    <ish-product-links-carousel
      [links]="links.other"
      [productLinkTitle]="'product.product_links.other.title' | translate"
    ></ish-product-links-carousel>
  </ng-container>
  <ng-container *ngIf="links.replacement?.products?.length">
    <ish-product-links-carousel
      [links]="links.replacement"
      [productLinkTitle]="'product.product_links.replacement.title' | translate"
    ></ish-product-links-carousel>
  </ng-container>
</ng-container>
