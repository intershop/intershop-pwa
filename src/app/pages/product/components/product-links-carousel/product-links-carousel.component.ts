import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { SwiperConfigInterface } from 'ngx-swiper-wrapper';

import { ProductLinkView } from 'ish-core/models/product-links/product-links.model';

/**
 * The Product Link Carousel Component
 *
 * Displays the products which are assigned to a specific product link type as an carousel.
 * It uses the {@link ProductItemContainerComponent} for the rendering of products.
 *
 * @example
 * <ish-product-links-carousel [links]="links.crossselling" [productLinkTitle]="'product.product_links.crossselling.title' | translate"></ish-product-links-carousel>
 */
@Component({
  selector: 'ish-product-links-carousel',
  templateUrl: './product-links-carousel.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ProductLinksCarouselComponent {
  /**
   * list of products which are assigned to the specific product link type
   */
  @Input() links: ProductLinkView;
  /**
   * title that should displayed for the specific product link type
   */
  @Input() productLinkTitle: string;
  /**
   * configuration for swiper carousel
   */
  @Input() swiperConfig: SwiperConfigInterface;
}
