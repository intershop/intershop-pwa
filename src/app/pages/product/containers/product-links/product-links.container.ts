import { ChangeDetectionStrategy, Component, Input, OnChanges } from '@angular/core';
import { Store, select } from '@ngrx/store';
import { Observable } from 'rxjs';

import { ProductLinksView } from 'ish-core/models/product-links/product-links.model';
import { LoadProductLinks, getProductLinks } from 'ish-core/store/shopping/products';

@Component({
  selector: 'ish-product-links-container',
  templateUrl: './product-links.container.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ProductLinksContainerComponent implements OnChanges {
  @Input() sku: string;

  links$: Observable<ProductLinksView>;

  constructor(private store: Store<{}>) {}

  ngOnChanges() {
    if (this.sku) {
      this.store.dispatch(new LoadProductLinks({ sku: this.sku }));
      this.links$ = this.store.pipe(select(getProductLinks, { sku: this.sku }));
    }
  }
}
