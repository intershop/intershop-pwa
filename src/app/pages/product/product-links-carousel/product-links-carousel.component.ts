import { ChangeDetectionStrategy, Component, Inject, Input } from '@angular/core';
import { SwiperOptions } from 'swiper';
import SwiperCore, { Navigation, Pagination } from 'swiper/core';

import { LARGE_BREAKPOINT_WIDTH, MEDIUM_BREAKPOINT_WIDTH } from 'ish-core/configurations/injection-keys';
import { ProductLinks } from 'ish-core/models/product-links/product-links.model';

SwiperCore.use([Navigation, Pagination]);

/**
 * The Product Link Carousel Component
 *
 * Displays the products which are assigned to a specific product link type as an carousel.
 * It uses the {@link ProductItemContainerComponent} for the rendering of products.
 *
 * @example
 * <ish-product-links-carousel [links]="links.crossselling" [productLinkTitle]="'product.product_links.crossselling.title' | translate"></ish-product-links-carousel>
 */
@Component({
  selector: 'ish-product-links-carousel',
  templateUrl: './product-links-carousel.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ProductLinksCarouselComponent {
  /**
   * list of products which are assigned to the specific product link type
   */
  @Input() links: ProductLinks;
  /**
   * title that should displayed for the specific product link type
   */
  @Input() productLinkTitle: string;

  /**
   * configuration of swiper carousel
   * find possible parameters here: http://idangero.us/swiper/api/#parameters
   */
  swiperConfig: SwiperOptions;

  constructor(
    @Inject(LARGE_BREAKPOINT_WIDTH) largeBreakpointWidth: number,
    @Inject(MEDIUM_BREAKPOINT_WIDTH) mediumBreakpointWidth: number
  ) {
    this.swiperConfig = {
      direction: 'horizontal',
      keyboard: true,
      mousewheel: false,
      navigation: true,
      scrollbar: false,
      breakpoints: {
        0: {
          slidesPerView: 2,
          slidesPerGroup: 2,
        },
        [mediumBreakpointWidth]: {
          slidesPerView: 3,
          slidesPerGroup: 3,
        },
        [largeBreakpointWidth]: {
          slidesPerView: 4,
          slidesPerGroup: 4,
        },
      },
      pagination: {
        el: '.swiper-pagination',
        clickable: true,
        clickableClass: 'swiper-pagination-clickable',
      },
    };
  }
}
