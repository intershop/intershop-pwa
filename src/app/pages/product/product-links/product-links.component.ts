import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';
import { Observable } from 'rxjs';

import { ProductContextFacade } from 'ish-core/facades/product-context.facade';
import { ProductLinksDictionary } from 'ish-core/models/product-links/product-links.model';

/**
 * The Product Link Container Component
 *
 * Collects all types of product links and prepares their rendering.
 * Product links can be displayed as product list (uses {@link ProductLinkListComponent}) or product carousel (uses  {@link ProductLinksCarouselComponent}).
 * For the carousel swiper is used.
 *
 * @example
 * <ish-product-links-list [links]="links.upselling" [productLinkTitle]="'product.product_links.upselling.title' | translate"></ish-product-links-list>
 */
@Component({
  selector: 'ish-product-links',
  templateUrl: './product-links.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ProductLinksComponent implements OnInit {
  links$: Observable<ProductLinksDictionary>;

  constructor(private context: ProductContextFacade) {}

  ngOnInit() {
    this.links$ = this.context.select('links');
  }
}
