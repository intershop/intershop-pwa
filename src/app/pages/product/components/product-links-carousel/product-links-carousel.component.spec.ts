import { ComponentFixture, TestBed, async } from '@angular/core/testing';
import { MockComponent } from 'ng-mocks';
import { SwiperModule } from 'ngx-swiper-wrapper';

import { ProductLinkView } from 'ish-core/models/product-links/product-links.model';

import { ProductItemContainerComponent } from '../../../../shared/product/containers/product-item/product-item.container';

import { ProductLinksCarouselComponent } from './product-links-carousel.component';

describe('Product Links Carousel Component', () => {
  let component: ProductLinksCarouselComponent;
  let fixture: ComponentFixture<ProductLinksCarouselComponent>;
  let element: HTMLElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [SwiperModule],
      declarations: [MockComponent(ProductItemContainerComponent), ProductLinksCarouselComponent],
    }).compileComponents();
  }));

  beforeEach(() => {
    const productLink = { productSKUs: ['sku'], categoryIds: ['catID'] } as ProductLinkView;

    fixture = TestBed.createComponent(ProductLinksCarouselComponent);
    component = fixture.componentInstance;
    element = fixture.nativeElement;
    component.links = productLink;
  });

  it('should be created', () => {
    expect(component).toBeTruthy();
    expect(element).toBeTruthy();
    expect(() => fixture.detectChanges()).not.toThrow();
    expect(element).toMatchSnapshot();
  });
});
