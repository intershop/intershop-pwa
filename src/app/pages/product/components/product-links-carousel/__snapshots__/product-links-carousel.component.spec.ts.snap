// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Links Carousel Component should be created 1`] = `
<div class="product-list-container">
  <h2></h2>
  <div class="product-list">
    <div class="product-list-item list-view">
      <ish-product-item-container
        displaytype="tile"
        ng-reflect-product-sku="sku"
        ng-reflect-display-type="tile"
      ></ish-product-item-container>
    </div>
  </div>
</div>
`;
